client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx

```diff
@@ -0,0 +1,150 @@
+import React, { useState, useEffect, useCallback, useRef } from 'react';
+import { Layout, Card, Steps, Space, Button, Typography } from 'antd';
+import HeaderNavbar from '../../content/HeaderNavbar/HeaderNavbar';
+import RobotInteractionInputSection from '../../content/RobotInteractionSections/RobotInteractionInputSection';
+import { getAllParametersForRobot } from '../../../api/variableRetrieval';
+import socket from '../../../utils/socket/socketConnections';
+import { upsert } from '../../../utils/attributeAndParamUtils';
+import styles from './RobotInteractionCockpit.module.css';
+import { isRobotExecutable } from '../../../utils/robotExecution';
+
+const { Step } = Steps;
+const { Title } = Typography;
+
+/**
+ * @description Page, where you can interact with a robot and for example enter input
+ * @category Client
+ * @component
+ */
+const RobotInteractionCockpit = (match) => {
+  const { robotId } = match.match.params;
+  const { userId } = match.location.state;
+  const isMounted = useRef(true);
+  const [parameterList, setParameterList] = useState([]);
+  const [currentStep, setCurrentStep] = useState(0);
+
+  /**
+   * @description For each activity of the current robot get the id, the name and all the parameters that require a user input
+   */
+  const getActivityAndParameterInformation = useCallback(() => {
+    getAllParametersForRobot(robotId)
+      .then((response) => response.json())
+      .then((parameterObjects) => {
+        const activityInformationList = [];
+        Array.prototype.forEach.call(parameterObjects, (parameterObject) => {
+          const { activityId } = parameterObject;
+          let activityName = '';
+          let ssot = sessionStorage.getItem('ssotLocal');
+          ssot = JSON.parse(ssot);
+          Array.prototype.forEach.call(ssot.elements, (elem) => {
+            if (elem.id === activityId) {
+              activityName = elem.name;
+            }
+          });
+          const activityParameter = [];
+          Array.prototype.forEach.call(
+            parameterObject.rpaParameters,
+            (parameter) => {
+              if (parameter.requireUserInput) {
+                activityParameter.push(parameter);
+              }
+            }
+          );
+          if (activityParameter.length !== 0) {
+            const activityInformation = {
+              activityId,
+              activityParameter,
+              activityName,
+            };
+            activityInformationList.push(activityInformation);
+          }
+        });
+        if (isMounted.current) {
+          setParameterList(activityInformationList);
+        }
+      });
+  }, [robotId]);
+
+  /**
+   * @description Equivalent to ComponentDidMount in class based components
+   */
+  useEffect(
+    () => () => {
```

Comment:
Why does this need to be a function that returns a function that does something?

Reply:
With _getActivityAndParameterInformation_ we are using an asynchronous callback that uses setState functionality (setParameterList). We want to ensure that this setState call is not executed when the component was already unmounted. Thats why we use _useRef_ and save us with isMounted if the component is mounted. When using useEffect without any dependencies like the useEffect here in line 71, we can return a function. This function will be saved by useEffect and executed when the component unmounts. In our case we are setting isMounted.current = true. And here I actually made a mistake. We should of course set isMounted.current = **false**. I just fixed that. ✅ With that we have ensured that setParameter will not be called when the component was unmounted.